### ISSUE TITLE
Add support for diff uploads

### ISSUE THREAD

**Issue by zerebubuth** [zerebubuth/openstreetmap-cgimap#140]
Now that we have [support for OAuth](https://github.com/zerebubuth/openstreetmap-cgimap/blob/e78578817f4b8f66dfc4e0b91ca1bcfe82cd047c/include/cgimap/oauth.hpp), we can add support for diff (i.e: `osmChange`) uploads. It seems like there's [a considerable improvement to be made by batching queries to the database](https://github.com/openstreetmap/openstreetmap-website/issues/1710#issuecomment-353774921).

Considerations:

* This will be cgimap's first write operation, so we should make sure we've got excellent test coverage and the code has been thoroughly reviewed to avoid corrupting the database.
* Although cgimap supports OAuth for viewing redacted versions of historical elements, it doesn't support HTTP Basic Auth. Only moderators use the redacted-history API, and they can probably live without HTTP Basic Auth. But there might be other people using Basic for the regular user API. Do we want to support that too? Or work with something like [cgimap-ruby](https://github.com/zerebubuth/cgimap-ruby) to allow the auth to be handled at a higher level?

**Comment by mmd-osm** [comment]
Moved previous comments about batching in this issue now:

---

I think it's time for a bit out of the box thinking, and leaving the current implementation behind for a minute. That's why I gave [pyosmium](http://osmcode.org/pyosmium/) and some bulk SQL statements a short try.

The same 8MB diff file was completely parsed in 140ms. Based on this I generated some bulk insert statements, effectively creating all 4300+ nodes in one go:

```sql
BEGIN;
INSERT INTO "current_nodes" ("latitude", "longitude", "changeset_id", "visible", "timestamp", "tile", "version") VALUES
(423712730, -832172277, 19, 't', '2017-12-24 09:15:27.204641', 1234, 1),
(423712670, -834339064, 19, 't', '2017-12-24 09:15:27.204675', 1234, 1),
(423715247, -832172360, 19, 't', '2017-12-24 09:15:27.204684', 1234, 1),
(423715140, -834339084, 19, 't', '2017-12-24 09:15:27.204691', 1234, 1),
(423717479, -834339184, 19, 't', '2017-12-24 09:15:27.204697', 1234, 1),
(423672264, -832171144, 19, 't', '2017-12-24 09:15:27.204704', 1234, 1),
(423672601, -834336964, 19, 't', '2017-12-24 09:15:27.204710', 1234, 1),
(423674831, -832171161, 19, 't', '2017-12-24 09:15:27.204717', 1234, 1)
[...many more...]
 RETURNING (id);
COMMIT;

-- Also possible as prepared statement:
-- prepare prep AS INSERT INTO "current_nodes" ("latitude", "longitude", "changeset_id", "visible", 
-- "timestamp", "tile", "version")  VALUES ($1, $2, $3, $4, $5, $6, $7), ($8, $9, $10, $11, $12, $13, $14) -- RETURNING(id);
-- execute prep(423712730, -832172277, 19, 't', '2017-12-24 09:15:27.204641', 1234, 1, 423712670, -834339064, 19, 't', '2017-12-24 09:15:27.204675', 1234, 1);

```

This query only took a few hundred ms on my local db (not much in there). The number of entries  (=package size) in one such INSERT statement could of course be made configurable, and the returned node ids used for later replacement operations. They could even be stored in a temporary table.

For deleting nodes (making them invisible), we could delegate most operations to the database:

```sql
 --- lock node ids first
select id from current_nodes where id in (1000,1001,1002) for update;

 --- check that all node/version pairs we want to update exist in current_nodes
 --- otherwise they already have been changed by another transaction
 --- and it's time to rollback this transaction  then (optimistic locking)
 --- For the sake of having proper error message with all conflicting pairs
 --- it would be good to get a list of all non-matching node id / version
 --- pairs (or at least one offending id/version pair by using LIMIT 1).

(select statement not shown here - should also be a bulk operation using temp table or VALUES)

 --- update current nodes for all node_ids: maybe via temp table containing node_id, version, 
 --- and new values for timestamp, visible, etc. possibly followed by a check that the updated row count is correct

(update statement not shown here. also either using temporary table or like illustrated here: https://stackoverflow.com/a/18799497/3501889)

--- create a copy of current nodes in the history table
insert into nodes (select id as node_id, latitude, longitude, 
          changeset_id, visible, timestamp, tile, version
          from current_nodes 
          where id in (1000,1001,1002));

--- since we deleted a bunch of nodes, we want to get rid of their tags, too.
delete current_node_tags where node_id in (1000,1001,1002);
```

For modify operations, a huge part of the task is to pull data out of the database, and INSERT it node by node, tag by tag in the history table. I wonder, why we don't let the database do this job and only provide a list of object ids to do the copying in one go. This takes maybe a second.

Here's a very rough idea. Most part of what I mentioned for DELETE before also applies here, I'm not repeating it here, though.


```sql
insert into node_tags ( select node_id, version, k, v from current_node_tags t
inner join current_nodes n on t.node_id = n.id where id in (1000,1001,1002));
```


I'm far away from understanding all of the details in the current implementation. Nevertheless, I think there's a lot of potential for improvement here, although it could mean that we can't reuse much of the current implementation.

**Comment by mmd-osm** [comment]
[WIP] Prototype implementation --> https://github.com/mmd-osm/animated-happiness

**Comment by pnorman** [comment]
> Only moderators use the redacted-history API, and they can probably live without HTTP Basic Auth.

I use HTTP Basic Auth over HTTPS most of the time when I need to use the redacted history API.

> But there might be other people using Basic for the regular user API. Do we want to support that too?

I expect there's a fair number of editors using that. Some even over HTTP.

Right now, for feature parity, we'd need to either keep HTTP Basic requests going to the rails port or support them in cgimap.

I wouldn't be opposed to dropping basic auth support, and would support dropping basic auth without HTTPS, but this issue tracker isn't the right place to make that policy change.

**Comment by mmd-osm** [comment]
A quick update from my side. I completed a first [proof of concept implementation](https://github.com/mmd-osm/animated-happiness), which covers adding/modifying/deleting nodes/ways/relations, including tags, way nodes and relation members. Also, copying from current_* tables to history tables is part of the PoC, as well as an update of the number of changes in the changeset.

A previous test case with 4241 nodes and some 180'000 tags in total that took around 15-20 minutes on rails now finishes in just 5 seconds. The code already survives a JOSM editing session as a [drop in replacement for rails code](https://github.com/mmd-osm/animated-happiness/wiki/Test).

In no particular order, I see a number of follow up activities:

- [ ] Review code / overall approach, changes to locking concept, ...

- [ ] Decide how to merge with cgimap

- [ ] Prepare cgimap handler for http post calls

- [ ] Possibly replace libosmium by own parser and osmobjects
   * Avoid extra lib
   * Solves issues with missing logic for create/modify/delete tags, which is currently based on version numbers. This doesn't really work all the time, in particular for create elements.
   * Also, `if-unused` attribute is not handled by libosmium

- [ ] Prepared statements 
  * Move to separate method, to be called only once
   * Use arrays everywhere, includes string[] and boolean[] which are not supported right now

- [ ] Adding proper validations, error messages and exceptions everywhere
   * Reject multiple versions of an element in osmchange file

- [ ] Implement flush limit for processing to limit memory consumption

- [ ] Update Changeset Bounding Box according to edits

- [ ] Adding testcases

**Comment by mmd-osm** [comment]
@pnorman : as GSoC 2018 is coming up soon, and the API topic is on the [proposed projects page](https://wiki.openstreetmap.org/wiki/Google_Summer_of_Code/2018/Project_Ideas#API) again, I think it would make sense to review the "Full cgimap write support" proposal, given that we probably want to incorporate the upload redesign in this issue in some way.

My other question would be if you (and possibly @Komzpa) could spend some time mentoring the upload redesign even outside of some GSoC framework, i.e. by reviewing code, bringing in ideas for improvement, providing guidance on how to add POST support to cgimap, which we could then route to my implementation.

There's still lots of work to be done to get test cases up and running, and any support here would be highly welcomed as well.

https://github.com/mmd-osm/diffupload has some details on how to get this code running as external binary called form Rails.

**Comment by mmd-osm** [comment]
@zerebubuth : I started integrating my coding into cgimap now, please see https://github.com/mmd-osm/openstreetmap-cgimap/tree/feature/bulk_upload

**Comment by mmd-osm** [comment]
After discussion with joto we came to the conclusion that libosmium might not be ideally suited for the task at hand. A [libosmium free implementation](https://github.com/mmd-osm/diffupload/tree/reorg2) is now waiting be integrated next into my cgimap branch.

@zerebubuth : as always, any kind of feedback is highly appreciated.

**Comment by Komzpa** [comment]
@mmd-osm it would be cool if you can get in touch with @hesidoryn who is sketching golang version in https://github.com/hesidoryn/fastmap-wrapper - it may be much easier to port it all to golang as is, since it already has rich HTTP API framework embedded.

**Comment by tomhughes** [comment]
Yes clearly this is the time to start over in a different language. Can we like finish this rewrite before starting on a new one please?

**Comment by Komzpa** [comment]
@thughes given current cgimap incarnation suffers from the same kind of architectural issues as railsport, it does not really matter which technology to use to start over as you'll have to start from scratch anyway.
There is no "finish" either. :)

**Comment by tomhughes** [comment]
What "architectural issues" are you talking about? The whole point of cgimap is to avoid the architectural problems of using rails, and as far as I'm aware it does so. That's not so say it can't be improved of course but the basic architecture is good.

Of course there is a "finish" as there are a finite number of API methods to implement and once they are all done the port away from rails is finished. Obviously more API methods might be added in future but that's the same regardless of technology.

The point is that given limited engineering resource it's insane to keep forking it rather than concentrating on getting one thing done. Unfortunately it's more exciting to start your own new thing in the latest fancy language than to contribute to an existing project.

**Comment by Komzpa** [comment]
@tomhughes I'm talking about issues described in detail in https://github.com/openstreetmap/operations/issues/135, that currently most of time of the api calls is lost in database roundtrips, on etiher rails or cgimap APIs. This behavior is dictated by "let's issue only selects without joins and sorts" concept in cgimap and the whole object model layout. There just isn't an idea that you may get all data you need (from all the tables you need) in a single database query in cgimap.

There is also no "engineering resource" at all. You're not paying anybody here to be in position to dehumanize them and call them "resource". People are volunteers and may do whatever they want. Only result you get from telling someone "stop doing it" is them possibly stop doing it, not switching the thing you may want them to.

If you think I shouldn't be getting people working on the similar topic in touch @tomhughes you should really reconsider what you're doing here.

**Comment by mmd-osm** [comment]
> I'm talking about issues described in detail in openstreetmap/operations#135, that currently most of time of the api calls is lost in database roundtrips, on etiher rails or cgimap APIs. This behavior is dictated by "let's issue only selects without joins and sorts" concept in cgimap and the whole object model layout. There just isn't an idea that you may get all data you need (from all the tables you need) in a single database query in cgimap.

That's long since implemented in cgimap and no longer the case.

**Comment by zerebubuth** [comment]
> currently most of time of the api calls is lost in database roundtrips, on etiher rails or cgimap APIs.

That's undoubtedly true of the diff upload API call, and possibly several others. I hope it's no longer true of anything that goes through cgimap, as we had some awesome contributions from @jronak in #130 and #136 which meant cgimap could make a single database call for extracting nodes, one for extracting ways and another for extracting relations.

No doubt it would be possible to get that down to a single call, but the extra saving of a handful of round-trips didn't seem to be worth the extra effort. According to https://github.com/zerebubuth/openstreetmap-cgimap/issues/122#issue-196250219, each round-trip adds about 0.2ms, so a few 10s of round-trips isn't too bad. We had issues when cgimap was making 1000s of round-trips for individual objects.

> This behavior is dictated by "let's issue only selects without joins and sorts" concept in cgimap and the whole object model layout.

It's not an explicit goal of cgimap to only issue selects without joins or sorts. In fact, cgimap now does [many joins](https://github.com/zerebubuth/openstreetmap-cgimap/blob/e78578817f4b8f66dfc4e0b91ca1bcfe82cd047c/src/backend/apidb/writeable_pgsql_selection.cpp#L475-L509), some of which require [sorting](https://github.com/zerebubuth/openstreetmap-cgimap/blob/e78578817f4b8f66dfc4e0b91ca1bcfe82cd047c/src/backend/apidb/writeable_pgsql_selection.cpp#L493). I've tried to avoid sorting on the whole query because I'm [worried about the extra load on the database](https://github.com/zerebubuth/openstreetmap-cgimap/issues/88). I'd love to be proved wrong about that, and we could make a few simple changes to enable sorted output.

> People are volunteers and may do whatever they want.

I agree totally, and I hope to see some awesome stuff from [fastmap-wrapper](https://github.com/hesidoryn/fastmap-wrapper) in the future. Having multiple projects means we can experiment with better ideas. If I were designing cgimap from scratch today then it would look totally different, and there's a number of mistakes in the design which have either required a lot of effort fix, or continue to require effort to work with.

For example, the backend "independence" isn't something cgimap really uses, and leads to [some ugly code](https://github.com/zerebubuth/openstreetmap-cgimap/blob/e78578817f4b8f66dfc4e0b91ca1bcfe82cd047c/src/api06/changeset_handler.cpp#L21-L23) to try and continue supporting.

The [data selection interface](https://github.com/zerebubuth/openstreetmap-cgimap/blob/e78578817f4b8f66dfc4e0b91ca1bcfe82cd047c/include/cgimap/data_selection.hpp) was intended to abstract over the data model, allowing handlers to be written in a style which concentrated on _what_ was being queried, rather than _how_ - ideally I wanted something which was close to a graph query like [Gremlin](https://en.wikipedia.org/wiki/Gremlin_(programming_language)), but unfortunately it didn't really work out that way.

For sure, there's a [lot of code in cgimap which ought to be provided by some HTTP standard library](https://github.com/zerebubuth/openstreetmap-cgimap/blob/e78578817f4b8f66dfc4e0b91ca1bcfe82cd047c/src/http.cpp), as it would be in Go and any other sensible modern language for writing web services! Unfortunately, the origins of cgimap date back to 2009, so it carries some baggage because of assumptions which were true but are no longer, and the languages and libraries which were available at the time.

@mmd-osm: Thanks so much for the work that you've been doing. I'm sorry I haven't had a chance to review it yet, but I'll try to get to that as soon as I can.

**Comment by zerebubuth** [comment]
@mmd-osm I had a quick look through and it looks good so far, thanks! Your code style is a little different from the repo style, but that's not worth worrying about now. The biggest issues seem to be plugging it into the request flow (i.e: not `demo()`), and the lack of tests.

I've started trying to add code so we can support POST requests. Will PR that into your branch as soon as it's ready.

**Comment by mmd-osm** [comment]
> Yes clearly this is the time to start over in a different language. Can we like finish this rewrite before starting on a new one please?

Now that the OSM API rewrite in golang is an official GSoC 2018 project, the question about priorities is becoming a bit more interesting now.

**Comment by tomhughes** [comment]
WTF

**Comment by mmd-osm** [comment]
Yes, seriously: https://wiki.openstreetmap.org/wiki/Google_Summer_of_Code/2018/AcceptedProjects

**Comment by tomhughes** [comment]
No mentor? But seriously, who was responsible for deciding that without any consultation?

**Comment by zerebubuth** [comment]
This is a cgimap issue thread. It's not an appropriate place to be discussing the approval process for GSoC applications.

While on a technical level I feel it's a shame to be duplicating work, it's clear that there's no great enthusiasm for the current cgimap implementation (even from me - my apologies for all the outstanding stuff blocked on me). If a new implementation is able to attract more sustained effort and maintenance, then that's a good thing for OSM. Even if the motivation for a new implementation is language chauvinism or other non-technical reasons.

**Comment by mmd-osm** [comment]
If I understood the application right, the scope of the GSoc project will cover the whole of OSM API 0.6, not just the parts that are currently implemented in cgimap. I don't want to go into the architectural implications on the Rails port (or projects like cgimap-ruby), but I feel that the scope is pretty huge for a 3 months project.

Which brings me back to this issue: I think it's pretty clear that porting the current Rails based diff upload with its many single selects is nothing we want to advocate for a future golang based implementation. On the other hand redesigning this thing took me several weeks to figure out how the Rails port works, and then redesign and implement it as bulk insert. The results are pretty amazing with more than factor 100 speedup compared to the current Rails based implementation.

I couldn't really figure out how to integrate the standalone piece into cgimap, that's where I heavily depend on @zerebubuth insight.  I already have a number of test cases already waiting here to be integrated, which I extracted from the current Rails port. It's just a matter of @zerebubuth setting up the environment to have 1 test case and I could continue with the rest.

My motivation is pretty clear: instead of making the job for the student project even harder, let's provide a rock solid basis for a C++ --> Golang port and try to do as much of the heavy lifting upfront as possible. I'm sure this will be a clear win win.

Let's jump right into it and at least pretend to be a bit excited... 😃

**Comment by pnorman** [comment]
> If I understood the application right, the scope of the GSoc project will cover the whole of OSM API 0.6, not just the parts that are currently implemented in cgimap. I don't want to go into the architectural implications on the Rails port (or projects like cgimap-ruby), but I feel that the scope is pretty huge for a 3 months project (which starts mid of May 2018 btw).

No, it's just what's implemented in cgimap. If there is time new stuff might get worked on, but the priority is the ability to switch the deployment from cgimap to the go port.

**Comment by mmd-osm** [comment]
@zerebubuth : I think there's some extremely good progress already done in your previous refactoring work, so thanks for that. I'm continuing in baby steps now with the integration in my branch: https://github.com/zerebubuth/openstreetmap-cgimap/compare/feature/bulk_upload...mmd-osm:feature/bulk_upload



* Moved temporary table creation to writeable_pgsql_selection
* Moved some parts of the "demo" code to changeset_upload_responder
* ApiDB_Node_Updater : Node_Updater inheritance also applied to Ways, Relations and Changesets
* XML parser: added chunk size, fixed bug with empty attributes

I'm still able to compile the code as is, but didn't try to run it. Currently, I see a few blocking issues, which I hope are not too silly or obvious.

* changeset_upload_responder still needs uid, and the actual payload.
* changeset_upload depends on Transaction_Manager, but it's not clear, how it should get this from the framework.
* changeset_upload_responder still uses ApiDB_*_Updater instance, which doesn't make a whole lot of sense in 0.6 code. 
* returning results: xml_formatter could see some extra methods, specifically for diffResults

**Comment by mmd-osm** [comment]
@tomhughes : at some point in the future we may want to deploy the changeset upload redesign on the dev instance to get some wider community exposure and identify bugs we couldn’t come up with in our own unit tests. Has cgimap ever been deployed on the dev instance before, alongside a working rails port installation, and some Apache / lighttpd to route requests to either one. How would you roughly assess the time and effort needed? TIA.

**Comment by mmd-osm** [comment]
@tomhughes : I have another short question on how to deploy cgimap on the dev instance.

Some older commits in the chef repo indicate that the dev instance is also managed by _chef_ and some PPA for cgimap would be needed for deployment. I don't seem to find any information describing how this should actually work. Do I have to go that route or would compiling the sources on the dev instance also be feasible?

The new changeset upload doesn't introduce any new dependencies except for a C++11 compiler. So whatever worked in the past to deploy cgimap should basically still be valid.

I have merged all 3 open PRs in this repo into https://github.com/mmd-osm/openstreetmap-cgimap/tree/demo  - this could be used for testing.

**Comment by tomhughes** [comment]
Ah I forgot cgimap was coming from a PPA now - it used to be build from source.

I don't really want to reintroduce building from source, so a PPA is probably best but it will need to be separate to our main one or the package will need to have a different name.

**Comment by mmd-osm** [comment]
Hmmm. ok, this is going to be a bit of a challenge then, as I can't find the PPA packaging instructions for CGImap anywhere.  The `/debian` directory which usually the place for this sort of information isn't part of this repo - (like in the following example: https://github.com/valhalla/packaging).

As the openstreetmap-cgimap - 0.6.0-1~xenial1 package installs both shared libs and the cgimap binary, it doesn't make too much sense recreating them from scratch.

So once again, I'm stuck and hope that @zerebubuth might have a bit of spare time to look into this.

**Comment by zerebubuth** [comment]
Yup, apologies - that's not documented anywhere. I think I was so exhausted by the time I finally got it working that I didn't write it down anywhere.

The distro-specific packaging is kept on distro-specific branches, of which there's currently `ubuntu/wily` and `ubuntu/xenial`. I think we can probably ignore the `ubuntu/wily` one now. I used [`git-buildpackage`](https://honk.sigxcpu.org/projects/git-buildpackage/manual-html/gbp.building.html) to manage the build process, as that seemed to be the least awful way of doing it.

The process of building is something like (apologies if this doesn't work - mostly from memory):

1. Tag release version on `master`.
2. Merge release version into `ubuntu/xenial`.
3. Update `debian/changelog` in the branch to have a release message. I haven't been particularly good about making these very meaningful, mostly out of frustration with having to keep trying again and again before it works. I don't understand the version numbering system (`-1~ubuntu1`) and sometimes I have to fiddle with it to get a usable build. Worth pointing out that the formatting of that file is _very_ sensitive and brittle. If you get it wrong, then either the Debian build tools or Launchpad will reject the package without a useful error message.
4. Commit the changes.
5. `mkdir -p ../build-area`
6. `DEBSIGN_KEYID=XXXXXXXX DEB_SIGN_KEYID=XXXXXXXX DEBSIGN_MAINT="Your Name <you@example.com>" gbp buildpackage --git-tag --git-export-dir=../build-area --git-export=ubuntu/xenial --git-debian-branch=ubuntu/xenial` (If you don't have `gbp` then `sudo apt install git-buildpackage dh-autoreconf`). Note that the package version is taken from `debian/changelog`, so unless you change and commit that file, you might find `gbp` does nothing or errors out. Use whichever key you've uploaded to Launchpad. I had real trouble getting it to use the right key, which is why I've probably overspecced the environment variables.
7. `cd ../build-area` and [use `dput` to upload the `.changes` file](https://help.launchpad.net/Packaging/PPA/Uploading).

Based on whether the package is formatted and signed correctly, Launchpad may or may not send you an email. If you don't get an email and the package doesn't show up in the build queue, it's worth checking the `debian/changelog` file.

**Comment by zerebubuth** [comment]
Ooops, looks like I had a whole bunch of local changes too. Pushed those to `ubuntu/xenial` now, so it should be up-to-date with `master`.

**Comment by mmd-osm** [comment]
Many thanks for your very detailed description of the build process. I totally agree it's really mind-boggling.

There was one issue I encountered on my forked cgimap repo. `gbp buildpackage` always wanted to use your upstream package as a baseline and apply my changes as patches to it. Somehow those upstream packages seem to lack source code (that's at least what the error messages  tell me), which makes the whole approach a bit pointless. I also tried to follow https://unix.stackexchange.com/questions/324680/how-to-apply-a-patch-in-a-debian-package - results were also inconclusive.

I continued with a bit of research and found the following approach. I'm not totally happy with the results yet, as it doesn't run through the build process locally. That's a huge pain on its own as any issues will only show up on the launchpad build process log half an hour later.

```
tar cvfj openstreetmap-cgimap_0.6.1.orig.tar.bz2 openstreetmap-cgimap --exclude .git
debuild -kXXX_my_signing_keyXXX -S -sa
dput -f ppa:mmd-osm/cgimap-test-001 openstreetmap-cgimap_0.6.1-1~xenial1_source.changes
```

After a few iterations launchpad finally produced an amd64 binary for a testing version named  `openstreetmap-cgimap 0.6.1-1~xenial1 `:

https://launchpad.net/~mmd-osm/+archive/ubuntu/cgimap-test-001 (still subject to change).

I added the PPA to my local box, installed openstreetmap-cgimap and ran a few test cases, which seemed to ok.

I still have to sort out some compilation issues for the i386 version (not that it would be hugely interesting), and change locale from en_gb.UTF8 to C.UTF-8 in the sources. Those should be fairly minor issues.

For better visibility into failing test cases, I'd also propose the following change to debian/rules:

```
override_dh_auto_test:
        VERBOSE=1 dh_auto_test
```

**Comment by mmd-osm** [comment]
@tomhughes : now that we have PPAs for both xenial and bionic, I wonder what the next step would be. Open an issue on the operations tracker, maybe?

**Comment by tomhughes** [comment]
We're likely to be extremely busy in the near future, and it's very much non-trivial to do, but open a ticket and I'll get to it when I can.

**Comment by mmd-osm** [comment]
> open a ticket and I'll get to it when I can.

Thanks, will do soon.

Meanwhile, I also set up a mini test drive on http://static.64.206.46.78.clients.your-server.de:31337/user/mmd2 so people can already start playing around with the new upload without impacting the dev instance.

To try it out create a new user, wait up to 5 min. for auto confirm process to run in background (no email confirm needed), and go ahead editing in Potlatch 2, iD or JOSM (see instructions on the user page).

**Comment by simonpoole** [comment]
@mmd-osm not getting very far with testing uploads right now (getting a 400 without any further messaging), maybe a short interactive session on irc would be the easiest to debug this.

**Comment by mmd-osm** [comment]
@simonpoole : many thanks for your feedback. Could you somehow extract the osmChange message you used for upload? Which editor did you for testing? We could arrange an irc session maybe later tonight.

**Comment by simonpoole** [comment]
@mmd-osm hmm seems to be some emulator induced brokenness that I've seen before, so likely false alarm for now.

**Comment by mmd-osm** [comment]
That's interesting, `PUT /api/0.6/changeset/create` is still being forwarded to the Rails port by lighttpd (which acts as a reverse proxy), and is not part of the changeset upload reimplementation. So this could be some kind of some lighttpd misconfiguration. I guess you don't encounter this issue when testing againsst dev.openstreetmap.org?

**Comment by simonpoole** [comment]
Works ok vs dev instance etc..

**Comment by mmd-osm** [comment]
@simonpoole : I feel like chasing chunked transfer encoding issues on lighttpd might not be worth it, so  I decided to set up an Apache instance on port 80 now, following the [Chef cookbook](https://github.com/openstreetmap/chef/blob/master/cookbooks/web/templates/default/apache.backend.erb). It serves as a reverse proxy for fcgi (like in the cookbook), but also forwards requests to the Rails port 3000. Unfortunately, I'm not really familiar with Passenger and don't know how to set that up properly. 

Could you please give this alternative URL a try: http://static.64.206.46.78.clients.your-server.de

**Comment by simonpoole** [comment]
Works now, thanks. A simple diff upload completes now and seems to do the "right" thing..... 

now on to test if you meticulously copied all the error messages :-)

**Comment by simonpoole** [comment]
Sigh the celebrations where a bit premature, with my standard code it now fails in the actual diff upload with the error __Http header 'Content-Length' missing__ (which is true :-)).
[borkedupload2.txt](https://github.com/zerebubuth/openstreetmap-cgimap/files/2167910/borkedupload2.txt)

**Comment by simonpoole** [comment]
There's no problem in testing with the "working" code, but the problem still has to be resolved: on the one hand OkHttp is "fairly" popular (as in the likely the most used HTTP library globally), on the other hand supporting HTTP 1.1 doesn't exactly fall in the newfangled stuff category.

**Comment by mmd-osm** [comment]
Strictly speaking, CGImap doesn't even talk HTTP but FastCGI protocol and Apache and its modules do all the translation between HTTP/* and FastCGI. Now the question boils down to how chunked encoding is getting mapped onto the FastCGI protocol.

I wonder if you could make your Vespucci APK file available for testing somehow. After installing the released version from Github, I figured out that I'm missing proper endpoint information for OAuth, presumably has to be added to https://github.com/MarcusWolschon/osmeditor4android/blob/master/src/main/res/values/apis.xml.

**Comment by mmd-osm** [comment]
@simonpoole : thanks a lot. I could install the apk now and already reproduce the "Content-length missing" error.

**Comment by mmd-osm** [comment]
@simonpoole : I have updated the logic to extract the POST payload a bit and could now successfully upload a change via Vespucci: http://static.64.206.46.78.clients.your-server.de/changeset/2000000073

At this time, this change is only available on port 80 in a locally patched version. I will update the PPA if there are no further issues with this change.

**Comment by mmd-osm** [comment]
@zerebubuth : now that GSoC is over (and we're still on cgimap), I'm wondering a bit what might be the best way to move ahead. Deploying cgimap on dev is already triggered in a separate operations issue, so that more people will have a chance for the second round of tests.

Then there's a quite large pull request, including lots test cases for a good overall code coverage.  Do you still plan to review this code in some way, or should we just merge it? I think it would be incredibly useful to have some more feedback on the code, although this will for sure be quite time consuming to go through. Thanks!

**Comment by mmd-osm** [comment]
Thanks to @tomhughes tireless work on the Chef repo, the changeset upload is now ready for general testing on the dev instance: https://upload.apis.dev.openstreetmap.org :1st_place_medal: 

JOSM API endpoint: https://upload.apis.dev.openstreetmap.org/api
Access token key: YaTwrUFPGMlskcqc0eLmyVUxHpGabM9UP94VjcNG
Access token secret: KzpJY8IyPfI3tl9YgWdfxSmdlasSUZQIn5BJVjPI

It supports both Basic auth and OAuth, as well as uncompressed and compressed messages for osmChange upload.

Examples: https://upload.apis.dev.openstreetmap.org/user/mmd2/history
